### PR TITLE
fix saving/loading for interactive_learning_approach

### LINKED ISSUE #382: test if saving/loading work with interactive learning approach (esp the online learning part)
in previous experiments, i noticed that they weren't working because the predicate classifiers weren't being saved correctly. need to dig into this

### PR DESCRIPTION
**Pull request by ronuchit**
@tomsilver this completes all the TODOs on the predicates refactoring doc! woohoo!!

Usage: after, e.g.:
`python src/main.py --env cover --approach interactive_learning --seed 0 --excluded_predicates HandEmpty,Holding --num_test_tasks 5`
you can do:
`python src/main.py --env cover --approach interactive_learning --seed 0 --excluded_predicates HandEmpty,Holding --num_test_tasks 5 --load_approach --load_data`
to feel like the flash

closes #382

### CODE REVIEW

File: src/approaches/interactive_learning_approach.py [hunk 1]
```diff
@@ -117,7 +121,13 @@ def _relearn_predicates_and_nsrts(
                 (self._predicates_to_learn - {pred}) | {new_pred}
 
         # Learn NSRTs via superclass
-        self._learn_nsrts(dataset.trajectories, online_learning_cycle)
+        self._learn_nsrts(self._dataset.trajectories, online_learning_cycle)
+
+        # Save the things we need other than the NSRTs, which were already
+        # saved in the above call to self._learn_nsrts()
+        save_path = utils.get_approach_save_path_str()
+        with open(f"{save_path}_{online_learning_cycle}.DATA", "wb") as f:
```
Comment (tomsilver) [inline src/approaches/interactive_learning_approach.py:129]: just to check, the reason this doesn't overwrite the file that's created by the parent class is that the suffixes are different (DATA here vs NSRTs there). right?

File: src/approaches/interactive_learning_approach.py [hunk 2]
```diff
@@ -117,7 +121,13 @@ def _relearn_predicates_and_nsrts(
                 (self._predicates_to_learn - {pred}) | {new_pred}
 
         # Learn NSRTs via superclass
-        self._learn_nsrts(dataset.trajectories, online_learning_cycle)
+        self._learn_nsrts(self._dataset.trajectories, online_learning_cycle)
+
+        # Save the things we need other than the NSRTs, which were already
```
Comment (tomsilver) [inline src/approaches/interactive_learning_approach.py:126]: I think we need one other thing too -- `self._best_score`. and in general, we'll need to save/load any class attributes that can change after initialization. (cc @amburger66 )

Comment (tomsilver) [reply src/approaches/interactive_learning_approach.py:126]: actually, best_score might get set correctly through the re-running of the interactions? but regardless, I think we should add it to the things we're saving and loading.

File: src/approaches/interactive_learning_approach.py [hunk 3]
```diff
@@ -117,7 +121,13 @@ def _relearn_predicates_and_nsrts(
...
+        # saved in the above call to self._learn_nsrts()
+        save_path = utils.get_approach_save_path_str()
+        with open(f"{save_path}_{online_learning_cycle}.DATA", "wb") as f:
+            pkl.dump((self._dataset, self._predicates_to_learn), f)
```
Comment (tomsilver) [inline src/approaches/interactive_learning_approach.py:130]: instead of a tuple, can we make this a dict with keys `dataset, predicates_to_learn`?

**Review by tomsilver — COMMENTED** [review]
thank you! yay for being on the last TODO

File: src/approaches/interactive_learning_approach.py (hunk 1 above)
Comment (ronuchit) [reply src/approaches/interactive_learning_approach.py:129]: right

**Review by tomsilver — APPROVED** [review]
lgtm thanks!

**Review by amburger66 — APPROVED** [review]
thanks!

**Pull request merged**